lib/libshare/smb.c

```diff
@@ -138,10 +138,14 @@ smb_retrieve_shares(void)
 
 			if (strcmp(key, "path") == 0)
 				path = dup_value;
-			if (strcmp(key, "comment") == 0)
+			else if (strcmp(key, "comment") == 0)
 				comment = dup_value;
-			if (strcmp(key, "guest_ok") == 0)
+			else if (strcmp(key, "guest_ok") == 0)
 				guest_ok = dup_value;
+			else {
+				free(dup_value);
+				dup_value = NULL;
+			}
 
 			if (path == NULL || comment == NULL || guest_ok == NULL)
 				continue; /* Incomplete share definition */
```

Comment:
The `smb_free()` function is 100% functionally equivalent to `free(3)`.  Long ago it used the be the case that it was unsafe to pass a NULL to free but that's no longer true for any semi-modern  implementation of libc.  From the man page:

```
       The free() function frees the memory space pointed  to  by  ptr,  which
       must  have  been  returned by a previous call to malloc(), calloc(), or
       realloc().  Otherwise, or if free(ptr) has already been called  before,
       undefined behavior occurs.  **If ptr is NULL, no operation is performed**.
```

> share_file_fp, there may be two "key" line, there will be a memory leak.

You can explain this further.

> Three: there are unified ZFS user mode code style.

I don't understand, exactly what is the style issue.